### DISCUSSION TITLE
Block roads from outside of Valhalla and OSM

### DISCUSSION THREAD

**Discussion by trasfract** [valhalla/valhalla#4856] [Ideas]
Hello,

I work on a software which uses OSRM for travel time estimations at different dates and with distinct calibration models. 

I am looking into the possibility to move away from OSRM, replacing it with Valhalla.
There are a few reasons which I believe are game-changers, one of which being the dynamic costing abilities that Valhalla includes.
I believe it should make it possible to address differences (calibration-wise) without having to create multiple instances of the travel time computer.

I believe that such a switch to Valhalla allows for new features to become available.
For a long time we have tried to come up with a way to allow our users to flag roads as blocked directly by our users. 
The process I have in mind is for
* users to describe blocked roads on a map (likely leading to some GeoJSON data that I'll store)
* our backend system to convert GeoJSON data into Valhalla edge ids
* our backend system to use these edge ids to have them blocked within Valhalla for route/matrix endpoints

I have considered two ways to make that work, but I'd like your opinion on how to best approach this:
* a custom dynamic costing implementation
* a data provisioning mimicking traffic data

Can you enlighten me as to which approach is more relevant / easier to implement?
Thank you!
Benoît

**Comment by kevinkreiser** [comment]
yep that's right you don't have to have multiple instances of the router and data to support multiple times of day.

we already support blocking roads at runtime as part of the request. You can send a list of `avoid_locations` which we will resolve to edges and the costing will dissallow them. Now if you have a huge list of these for the whole world it doesn't make sense to send them all. But you could send just the ones that are within a certain bounds around the locations you are routing (make a bounding box with some padding around the input locations).

if your data is going to be quite dense or the requester doesn't have the data to send in the request or it changes a lot yes you might want to use the traffic data support.

loosely the process is

1. figure out how to match your geojson to the graph to get them converted to graph ids (locate api can do this)
2. create/update traffic tiles
3. push them into the traffic tile tar by overwriting the byte range where a given tile occurs

if the router is configured to use the traffic tar you should be good to go

**Reply by trasfract** [reply]
Thank you so much for your answer. Indeed, it looks I won't have much coding to do on Valhalla's side. The `avoid_locations` parameter is exactly the one I was looking for.

Based on what I have seen, I don't need more to properly block roads.
I don't think we'll get to a long list, but I'll keep your advice in mind in case we do.

Regarding the loose process you describe at the end of your message, it looks like it could be useful to calibrate traffic in our maps the way we are used to, with GeoJSON describing zones within which the traffic time travel estimates are to be multiplied by a specific factor. 
Are there specific bits of documentation that I should look into for that purpose? I have yet to read detailed specs on how to integrate traffic tile tar data. I may have failed to properly search your documentation (in which case I am sorry!).

Again thanks for the good pieces of advice!

**Reply by kevinkreiser** [reply]
it is not yet well documented and that is something ive owed to many other folks over the years and just have put off writing. its hard to find the time. i'll get to it at some point. at the moment the code and tests has to be the docs :frowning_face:

**Reply by trasfract** [reply]
I completely understand!
I have tried to find information elsewhere, and the closest I have found is this page: https://gis-ops.com/traffic-in-valhalla/
It is not clear enough for me to implement, but unfortunately, if what it says is true - that the `/matrix` endpoint does not work with traffic - then it won't work in our case.

What we are trying to provision Valhalla with is the following (example):
![image](https://github.com/user-attachments/assets/814ee54d-238c-4076-8bf0-cc5295591ca8)
The ratio would be used on top of the otherwise-computed speed for each edge.
And we'd like for it to be available for the `/matrix` endpoint, otherwise it becomes useless :-| 

Any thoughts on how to achieve this?

Thanks again for your help and your work on Valhalla which looks awesome!

**Reply by trasfract** [reply]
As an alternative, if I were to find a way to introduce traffic data into Valhalla, how much work would be needed to use this data for the `/matrix` endpoint?

**Reply by ianthetechie** [reply]
For some more traffic docs, have a look at this repo: https://github.com/alinmindroc/valhalla_traffic_poc.

I don't have an answer on traffic integration difficulty for matrix, other than that AFAIK it is not currently supported (at least not live traffic; historical might be?).

**Reply by trasfract** [reply]
Thank you for your help, I'll look into the contents of the above link.

In the meantime, I am quite curious as to what altered behavior one can expect from the `/matrix` endpoint with a provided datetime, as described here: https://valhalla.github.io/valhalla/api/matrix/api-reference/#costing-parameters
if it does not take into account any kind of traffic information.
Do you know what's taken into account in this case?

**Reply by trasfract** [reply]
For future readers: the PR detailing the addition of datetime to the matrix endpoint is located here: https://github.com/valhalla/valhalla/pull/3795/files
I am in the process of reading its contents :-)

**Reply by kevinkreiser** [reply]
you can also have a look at this one (for the other faster matrix algorithm): https://github.com/valhalla/valhalla/pull/4071
